deepinv/loss/measplit.py

```diff
@@ -121,42 +139,84 @@ def adapt_model(self, model, MC_samples=5):
             return SplittingModel(
-                model, self.split_ratio, self.mask_generator, MC_samples
+                model,
+                split_ratio=self.split_ratio,
+                mask_generator=self.mask_generator,
+                MC_samples=MC_samples,
+                eval_split_output=self.eval_split_output,
             )
 
 
 class SplittingModel(torch.nn.Module):
-    def __init__(self, model, split_ratio, mask_generator, MC_samples):
+    """
+    Model wrapper when using SplittingLoss.
+
+    Performs input splitting during forward pass. At evaluation, perform forward passes for multiple realisations of splitting mask and average.
+
+    For parameters, see :class:`deepinv.loss.SplittingLoss`
+    """
+
+    def __init__(
+        self, model, split_ratio, mask_generator, MC_samples, eval_split_output
+    ):
         super().__init__()
         self.model = model
         self.split_ratio = split_ratio
         self.MC_samples = MC_samples
         self.mask = 0
         self.mask_generator = mask_generator
+        self.eval_split_output = eval_split_output
 
     def forward(self, y, physics, update_parameters=False):
         MC = 1 if self.training else self.MC_samples
         tsize = y.size()[1:]
         out = 0
 
         if self.mask_generator is None:
-            self.mask_generator = BernoulliMaskGenerator(
+            self.mask_generator = BernoulliSplittingMaskGenerator(
                 tensor_size=tsize, split_ratio=self.split_ratio, device=y.device
             )
 
-        inp = Inpainting(
-            tsize, mask=0.0, device=y.device, noise_model=physics.noise_model
+        inp = Inpainting(tsize, mask=0.0, device=y.device)
+
+        eval_split_output = (
+            self.eval_split_output
+            and isinstance(physics, Denoising)
+            and not self.training
         )
+
         with torch.set_grad_enabled(self.training):
+
+            normaliser = torch.zeros_like(y)
+
             for i in range(MC):
-                mask = self.mask_generator.step(y.size(0))
+                mask = self.mask_generator.step(
+                    y.size(0),
+                    input_mask=None if not hasattr(physics, "mask") else physics.mask,
+                )
                 y1 = inp.A(y, **mask)
                 physics1 = inp * physics  # A_1 = P*A
-                out += self.model(y1, physics1)
+                physics1.noise_model = physics.noise_model
+                x_hat = self.model(y1, physics1)
+
+                if eval_split_output:
+                    mask2 = (
+                        1.0 if not hasattr(physics, "mask") else physics.mask
+                    ) - mask["mask"]
+                    inp2 = Inpainting(tsize, mask=mask2, device=y.device)
+                    x_hat = inp2.A(x_hat)
+                    normaliser += mask2
+
+                out += x_hat
+
+            if eval_split_output:
+                out[normaliser != 0] /= normaliser[normaliser != 0]

```

Comment:
if normaliser == 0, it means that out will have a hole...
we should raise some warning at least I guess

Reply:
What do you mean? Wherever normaliser==0, out also ==0 so the output will just be 0.

Reply:
yes this is what I meant - if the output is zero it will be probably be a bad output